File: leetcode/python/1396_designUndergroundSystem.py
import argparse
import collections
from typing import Dict, List, Tuple


class UndergroundSystem:
    def __init__(self):
        # (start, end): [totalTime, samples]
        self.station_times: Dict[Tuple[str, str], List[int, int]] = collections.defaultdict(
            lambda: [0, 0])
        # id: (stationName, checkInTime)
        self.current_passengers: Dict[int, Tuple[str, int]] = dict()

    def checkIn(self, id: int, stationName: str, t: int) -> None:
        if id in self.current_passengers:
            return
        self.current_passengers[id] = (stationName, t)

    def checkOut(self, id: int, stationName: str, t: int) -> None:
        if id not in self.current_passengers:
            return
        startStation, checkInTime = self.current_passengers[id]

        self.current_passengers.pop(id)

        self.station_times[(startStation, stationName)][0] += t-checkInTime
        self.station_times[(startStation, stationName)][1] += 1

    def getAverageTime(self, startStation: str, endStation: str) -> float:
        totalTime, samples = self.station_times[(startStation, endStation)]
        if samples == 0:
            return 0
        return totalTime/samples


# Your UndergroundSystem object will be instantiated and called as such:
# obj = UndergroundSystem()
# obj.checkIn(id,stationName,t)
# obj.checkOut(id,stationName,t)
# param_3 = obj.getAverageTime(startStation,endStation)

def test1():
    undergroundSystem = UndergroundSystem()

    # Customer 10 "Leyton" -> "Paradise" in 8-3 = 5
    undergroundSystem.checkIn(10, "Leyton", 3)
    undergroundSystem.checkOut(10, "Paradise", 8)

    # return 5.00000, (5) / 1 = 5
    r1 = undergroundSystem.getAverageTime("Leyton", "Paradise")
    assert r1 == 5

    # Customer 5 "Leyton" -> "Paradise" in 16-10 = 6
    undergroundSystem.checkIn(5, "Leyton", 10)
    undergroundSystem.checkOut(5, "Paradise", 16)

    # return 5.50000, (5 + 6) / 2 = 5.5
    r2 = undergroundSystem.getAverageTime("Leyton", "Paradise")
    assert r2 == 5.5

    # Customer 2 "Leyton" -> "Paradise" in 30-21 = 9
    undergroundSystem.checkIn(2, "Leyton", 21)
    undergroundSystem.checkOut(2, "Paradise", 30)

    # return 6.66667, (5 + 6 + 9) / 3 = 6.66667
    r3 = undergroundSystem.getAverageTime("Leyton", "Paradise")
    assert r3 == (5 + 6 + 9) / 3

    print(r1, r2, r3)


args_parser = argparse.ArgumentParser()
args_parser.add_argument('--test',  action='store_true',
                         default=False, required=False)

args = args_parser.parse_args()

if args.test:
    test1()

File: leetcode/rust/src/problems/_1396_design_underground_system.rs
#![allow(unused)]

use std::collections::HashMap;

struct UndergroundSystem {
    station_times: HashMap<(String, String), (f64, i32)>,
    current_passengers: HashMap<i32, (String, i32)>,
}

/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl UndergroundSystem {
    fn new() -> Self {
        return Self {
            station_times: HashMap::new(),
            current_passengers: HashMap::new(),
        };
    }

    // O(1)
    fn check_in(&mut self, id: i32, station_name: String, t: i32) {
        self.current_passengers.insert(id, (station_name, t));
    }

    // O(1)
    fn check_out(&mut self, id: i32, station_name: String, t: i32) {
        if let Some((entry_station, entry_time)) = self.current_passengers.get(&id) {
            let (ave_time, samples) = self
                .station_times
                .entry((entry_station.to_string(), station_name))
                .or_default();

            // Calculate rolling average

            // prev_ave * samples
            let curr_ave = *ave_time * (*samples as f64);
            // increment samples
            *samples += 1;

            // new ave = ((Prev*Count) + NewVal) / (Count+1)
            *ave_time = (curr_ave + (t - entry_time) as f64) / (*samples as f64);
        }
    }

    // O(1)
    fn get_average_time(&self, start_station: String, end_station: String) -> f64 {
        if let Some((ave_time, _)) = self.station_times.get(&(start_station, end_station)) {
            return *ave_time;
        }
        0.0
    }
}

//  Your UndergroundSystem object will be instantiated and called as such:
//  let obj = UndergroundSystem::new();
//  obj.check_in(id, stationName, t);
//  obj.check_out(id, stationName, t);
//  let ret_3: f64 = obj.get_average_time(startStation, endStation);

#[test]
fn test() {
    let mut underground_system: UndergroundSystem = UndergroundSystem::new();
    underground_system.check_in(10, String::from("Leyton"), 3);
    underground_system.check_out(10, String::from("Paradise"), 8); // Customer 10 String::from("Leyton") -> String::from("Paradise") in 8-3 = 5

    let r1 = underground_system.get_average_time(String::from("Leyton"), String::from("Paradise")); // return 5.00000, (5) / 1 = 5
    assert_eq!(r1, 5.0);

    underground_system.check_in(5, String::from("Leyton"), 10);
    underground_system.check_out(5, String::from("Paradise"), 16); // Customer 5 String::from("Leyton") -> String::from("Paradise") in 16-10 = 6

    let r2 = underground_system.get_average_time(String::from("Leyton"), String::from("Paradise")); // return 5.50000, (5 + 6) / 2 = 5.5
    assert_eq!(r2, 5.5);

    underground_system.check_in(2, String::from("Leyton"), 21);
    underground_system.check_out(2, String::from("Paradise"), 30); // Customer 2 String::from("Leyton") -> String::from("Paradise") in 30-21 = 9
    let r3 = underground_system.get_average_time(String::from("Leyton"), String::from("Paradise"));
    assert_eq!(r3, 20.0 / 3.0); // return 6.66667, (5 + 6 + 9) / 3 = 6.66667
}
